File: 32. Longest Valid Parentheses.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        stk = []
        ans = accumulate = 0
        
        for c in s:
            if c == '(':
                stk.append(0)
            
            elif len(stk)>1:
                stk[len(stk)-2] += (stk.pop()+2)
            
            elif stk:
                accumulate += (stk.pop()+2)
            
            else:
                ans, accumulate = max(ans, accumulate), 0
            
        stk.append(accumulate)
            
        return max(ans, max(stk))

File: 32. Longest Valid Parentheses.js
/**
 * @param {string} s
 * @return {number}
 */
var longestValidParentheses = function(s) {
    let stk = [];
    let accumulator = 0;
    let ans = 0;
    
    for (let i=0; i<s.length; i++){
        if (s[i]==='('){
            stk.push(0);
        }
        else if (stk.length>1){
            stk[stk.length-2] += (stk.pop()+2);
        }
        else if (stk.length==1){
            accumulator += (stk.pop()+2);   
        }
        else{
            ans = Math.max(ans, accumulator);
            accumulator = 0
        }
    }
    stk.push(accumulator);
    stk.push(ans);
    
    return Math.max(...stk);
};